Run Fastfile and Matchfile past Rubocop

mokagio opened the pull request:
Between my work to migrate secrets outside of the repo and my upcoming rotation as release manager, I expect to be touching the `Fastfile` a fair bit.

My Vim setup bombarded me with warnings on Ruby style used in the file, so I thought I'd take the time to address them.

Now, you could argue that I should have disabled the warnings on my end, but I'd say that, because the `Fastfile` _is_ a Ruby file, using a standard style (standard according to Rubocop, at least) is appropriate.

To test:

I struggle to come up with an efficient way to test these changes. Given they're all coding style related, and that the majority of them was automated via `rubocop -a` (see the first commit, 9ef77b4) reviewing this commit-by-commit should suffice.

## Regression Notes
1. Potential unintended areas of impact
N.A. See testing details above.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N.A.

3. What automated tests I added (or what prevented me from doing so)
N.A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

mokagio committed:
Run Fastfile under autocorrect

With `rubocop -a`

mokagio committed:
Remove trailing '/' from PROJECT_ROOT_FOLDER env assignment

It's unnecessary. All the logic in the release toolkit uses Ruby's File
API to build paths on top of that, which handles the path composition
already.

mokagio committed:
Address frozen strings and constants Rubocop violations in Fastfile

mokagio committed:
Use using snake_casing for all variables in Fastfile

mokagio committed:
Refactor missing env file message in Fastfile

This came from a line length Rubocop warning. While I decided to disable
the warning for Fastfile, I still think this update is useful from a DRY
point of view.

mokagio committed:
Address an idiomatic Ruby suggestion by Rubocop in Fastfile

mokagio committed:
Add .rubocop.yml to disable some cops for Fastfile

mokagio commented on fastlane/Fastfile:
```diff
@@ -49,7 +50,9 @@ before_all do |lane|
   unless lane == :test_without_building
     # Check that the env files exist
     unless is_ci || File.file?(USER_ENV_FILE_PATH)
-      UI.user_error!("~/.wpios-env.default not found: Please copy fastlane/env/user.env-example to #{USER_ENV_FILE_PATH} and fill in the values")
+      example_path = 'fastlane/env/user.env-example '
+      msg = "#{ENV_FILE_NAME} not found: Please copy #{example_path} to #{USER_ENV_FILE_PATH} and fill in the values"
+      UI.user_error!(msg)
```
I did this before disabling the line length violation. Still, it DRYes a bit (see `ENV_FILE_NAME`) and I don't think it hurst the code, so I decide to leave it.

jkmassel replied:
I'm not sure we need to disable the line length violation – splitting method calls onto multiple lines does improve readability anyway 🤔 

WDYT?

mokagio replied:
Yeah true. I mainly disabled it because of long string lines like https://github.com/wordpress-mobile/wordpress-ios/blob/fastlane-rubocop/fastlane/Fastfile#L419.

There's probably a Ruby way to split the string without messing up the horizontal line length 🤔

mokagio commented on fastlane/Fastfile:
```diff
@@ -778,7 +778,7 @@ platform :ios do
     end
 
-    UI.user_error!("Can't find any reference to Gutenberg!") unless res.length != 0
+    UI.user_error!("Can't find any reference to Gutenberg!") if res.empty?
```
First Rubocop suggested to use `res.empty? == false` instead of `res.length != 0`, then to flip `unless res.empty? == false` to `if res.empty?`.

I'm spelling this out to make sure it holds logically, as I'm painfully aware that meddling with if statements can backfire 😅

jkmassel replied:
It reads that it's doing the same thing to me!

mokagio committed:
Run Matchfile through Rubocop with autocorrect

mokagio committed:
Address frozen string literal and file name Matchfile Rubocop violations

mokagio commented on .rubocop.yml:
```diff
@@ -0,0 +1,15 @@
+
+Naming/FileName:
+  Exclude:
+    - fastlane/Matchfile
```
I can only guess that `Fastfile` is already excluded by the rule, while `Matchfile` isn't. 

I'd love to look into it on the Rubocop end and see if I can address it.

jkmassel replied:
IMHO we should just get rid of the `Matchfile` anyway, so that would also be a solution 😁

mokagio replied:
+1.

Maybe something I'll get to do over the next few weeks. Together with folding the `Deliverfile` and standalone `deliver` call into a lane to make it consistent with the rest of the workflow.

jkmassel commented on fastlane/Fastfile:
```diff
@@ -63,27 +66,27 @@ before_all do |lane|
+  ENV['PUBLIC_CONFIG_FILE'] = File.join(PROJECT_ROOT_FOLDER, 'config', 'Version.Public.xcconfig')
+  ENV['INTERNAL_CONFIG_FILE'] = File.join(PROJECT_ROOT_FOLDER, 'config', 'Version.internal.xcconfig')
+  ENV['DOWNLOAD_METADATA'] = './fastlane/download_metadata.swift'
+  ENV['PROJECT_ROOT_FOLDER'] = PROJECT_ROOT_FOLDER
```
I think this change might break things (internally IIRC there are assumptions that the `PROJECT_ROOT_FOLDER` env var ends in a trailing slash).

mokagio replied:
Good point. When I made this change in isolation in 52ee45a, I [grepped for `PROJECT_ROOT_FOLDER`](https://github.com/wordpress-mobile/release-toolkit/search?q=PROJECT_ROOT_FOLDER&type=).

```
➜ g grep PROJECT_ROOT
lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_download_translations_action.rb:        res_dir = File.join(ENV['PROJECT_ROOT_FOLDER'] || '.', params[:res_dir])
lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_update_release_notes.rb:        path = File.join(ENV['PROJECT_ROOT_FOLDER'] || '.', 'RELEASE-NOTES.txt')
lib/fastlane/plugin/wpmreleasetoolkit/actions/ios/ios_update_release_notes.rb:        path = File.join(ENV['PROJECT_ROOT_FOLDER'] || '.', 'RELEASE-NOTES.txt')
lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_git_helper.rb:        # @env PROJECT_ROOT_FOLDER The path to the git root of the project
lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_git_helper.rb:            files: File.join(ENV['PROJECT_ROOT_FOLDER'], ENV['PROJECT_NAME'], 'build.gradle'),
lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_git_helper.rb:        # @env PROJECT_ROOT_FOLDER The path to the git root of the project
lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_git_helper.rb:          res_dir = File.join(ENV['PROJECT_ROOT_FOLDER'], ENV['PROJECT_NAME'], 'src', 'main', 'res')
lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_version_helper.rb:          gradle_file_path = File.join(ENV['PROJECT_ROOT_FOLDER'] || '.', 'build.gradle')
lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_version_helper.rb:        # @env PROJECT_ROOT_FOLDER The path to the root of the project (the folder containing the `.git
lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_version_helper.rb:          UI.user_error!("You need to set the \`PROJECT_ROOT_FOLDER\` environment variable to the path
lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_version_helper.rb:          File.join(ENV['PROJECT_ROOT_FOLDER'], ENV['PROJECT_NAME'], 'build.gradle')
lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_git_helper.rb:        # @env PROJECT_ROOT_FOLDER The path to the git root of the project
lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_git_helper.rb:          files_list = [File.join(ENV['PROJECT_ROOT_FOLDER'], 'config', '.')]
lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_git_helper.rb:            files_list.append File.join(ENV['PROJECT_ROOT_FOLDER'], ENV['PROJECT_NAME'], 'Resources', ENV['APP_STOR
lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_git_helper.rb:        # @env PROJECT_ROOT_FOLDER The path to the git root of the project
lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_git_helper.rb:          Action.sh("cd #{get_from_env!(key: 'PROJECT_ROOT_FOLDER')} && ./Scripts/localize.py")
lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_git_helper.rb:        # @env PROJECT_ROOT_FOLDER The path to the git root of the project
lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_git_helper.rb:          Action.sh("cd #{get_from_env!(key: 'PROJECT_ROOT_FOLDER')} && ./Scripts/update-translations.rb")
lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_git_helper.rb:          project_root = get_from_env!(key: 'PROJECT_ROOT_FOLDER')
lines 1-19/19 (END)
```

AFAICS, all the instances are used within `File.join` which handles the `/`. That includes the [assignment to `project_root`](https://github.com/wordpress-mobile/release-toolkit/blob/develop/lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_git_helper.rb#L65-L70) in the last line of the `git grep` output.

Can you see anything that I missed? Otherwise, how do you feel about risking it and see if it breaks anything? (I could do it once I roll over as release manager)

I'm saying it mainly because removing the `/` is the right thing to do™

jkmassel replied:
> how do you feel about risking it and see if it breaks anything

I'd be ok with it, but then we're committing to a major version bump of this library – do we want that?

IMHO, while it is more correct to adjust this, I'd prefer to make a larger change to not require these global variables anymore (even if it's "not right" for a while).

If you're ok with bumping the major version we can do the change though 😃

jkmassel replied:
IMHO it's a cosmetic issue, so while it's not correct, I'd _personally_ prefer to address the structural problem (global variables) in the release toolkit? I'd be concerned about the potential for things silently breaking during a transition.

That said, you're the one who'd have to catch them, not me, so maybe just give it a think and feel free to go either way? 🙂

mokagio replied:
I was tempted to say "Yeah no worries. I checked the code and I'm sure that won't be any problems." Then I had visions of broken code freezes and tears of frustration.

Reverted! fb23430 (🙌 past me for small atomic commits)

Thank you for being wise.

jkmassel commented on fastlane/Fastfile:
```diff
@@ -330,15 +335,16 @@ import "./ScreenshotFastfile"
-      external: true)
-    ios_build_preflight()
+                        internal: options[:beta_release],
+                        external: true)
```
Nit: I find these multi-line method calls without the closing brace on its own line easy to break (and messy in diffs if we need to add/remove arguments).

Is putting the closing branch on its own line rubocop-compliant?

mokagio replied:
Totally!

I'm guessing it's an opt-in kind of setting? Otherwise it would have told me about it I'd say

jkmassel commented on fastlane/Fastfile:
```diff
@@ -533,58 +538,59 @@ import "./ScreenshotFastfile"
+    ios_merge_translators_strings(strings_folder: File.join(PROJECT_ROOT_FOLDER, 'WordPress', 'Resources'))
 
-    extract_framework_translations_script_path = File.join( File.join( PROJECT_ROOT_FOLDER, "Scripts", "extract-framework-translations.swift"))
+    extract_framework_translations_script_path = File.join(File.join(PROJECT_ROOT_FOLDER, 'Scripts',
```
If we're here anyway – do we need the nested `File.join`?

mokagio replied:
😅 good point

jkmassel reviewed (changes requested):
I saw one potential breaking change, but in general this looks solid, and I like the idea of maintaining according to a standard set of rules!

mokagio committed:
Remove unnecessary nested File.join from Fastfile

Nice catch @jkmassel

See
https://github.com/wordpress-mobile/WordPress-iOS/pull/16621#discussion_r645839262

mokagio commented:
@jkmassel I requested a new review as per my comment [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/16621#discussion_r647111063).

jkmassel reviewed (approved):
Approved to avoid blocking, but left some thoughts.

mokagio committed:
Revert "Remove trailing '/' from PROJECT_ROOT_FOLDER env assignment"

This reverts commit f41b4b794d22926cbdc078e23aefad84691067a3.

Rationale: Better safe but with a warning than sorry with broken automation!

See conversation with @jkmassel in
https://github.com/wordpress-mobile/WordPress-iOS/pull/16621/commits/b2c57f2d00f8c82577bdb802ca43fdb9ade100b2#r645838433
